src/zcash/prf.cpp

```diff
@@ -0,0 +1,63 @@
+#include "prf.h"
+#include "crypto/sha256.h"
+
+uint256 PRF(bool a, bool b, bool c, bool d,
+            const uint256& x,
+            const uint256& y)
+{
+    uint256 res;
+    unsigned char blob[64];
+
+    memcpy(&blob[0], x.begin(), 32);
+    memcpy(&blob[32], y.begin(), 32);
+
+    blob[0] &= 0x0F;
+    blob[0] |= (a ? 1 << 7 : 0) | (b ? 1 << 6 : 0) | (c ? 1 << 5 : 0) | (d ? 1 << 4 : 0);
+
+    CSHA256 hasher;
+    hasher.Write(blob, 64);
+    hasher.FinalizeNoPadding(res.begin());
+
+    return res;
+}
+
+uint256 PRF_addr(const uint256& a_sk, unsigned char t)
+{
+    uint256 y;
+    *(y.begin()) = t;
+
+    return PRF(0, 0, 0, 0, a_sk, y);
```

Comment:
The domain separation on this PRF is wrong, it should be **1, 1, 0, 0** as per the spec.